File: Utils/TemplateDiagnostics.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace NotifyMessages;

internal enum TemplateSeverity
{
    Error,
    Warning
}

/// Одна претензия к шаблону: что не так и где именно.
internal sealed record TemplateIssue(TemplateSeverity Severity, string Where, string Tag, string Text)
{
    public override string ToString()
    {
        var mark = Severity == TemplateSeverity.Error ? "ОШИБКА" : "внимание";
        return $"[{mark}] {Where}: {Text}";
    }
}

/// Статический анализатор шаблонов сообщений.
///
/// Смысл: тег, который не сможет подставить ни MessageProcessor, ни TextFormatter,
/// доезжает до игрока литеральным текстом в фигурных скобках. Раньше это находил
/// только игрок. Здесь это находит машина — при загрузке конфига и по команде.
///
/// Нативов внутри нет вообще: класс чистый и покрыт юнит-тестами.
internal static class TemplateDiagnostics
{
    // Больше трёх уровней вложенности ({prefix} -> текст -> ещё ключ) не бывает,
    // а ограничение заодно страхует от циклов вида {a} -> {b} -> {a}.
    private const int MaxDepth = 3;

    /// Теги, работающие лишь в отдельных местах конфига: значение подставляет вызывающий код,
    /// а не MessageProcessor. Значение словаря — человекочитаемое «где именно».
    internal static readonly IReadOnlyDictionary<string, string> ContextualTags =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["{PLAYERNAME}"] = "WelcomeMessage, ChangeTeamMessage, JoinTeamMessage, Join/LeaveMessages",
            ["{TEAM}"] = "ChangeTeamMessage, JoinTeamMessage",
            ["{OLD_TEAM}"] = "ChangeTeamMessage",
            ["{TIME_RESTART}"] = "RestartNotify",
            ["{SECONDS}"] = "RestartNotify",
            ["{COUNTRY}"] = "JoinMessages, LeaveMessages",
            ["{CITY}"] = "JoinMessages, LeaveMessages",
            ["{SERVER_IP}"] = "Servers.json",
            ["{SERVER_PORT}"] = "Servers.json",
            ["{SERVER_MAP}"] = "Servers.json",
            ["{SERVER_PLAYERS}"] = "Servers.json",
            ["{SERVER_MAXPLAYERS}"] = "Servers.json"
        };

    /// Разбирает один шаблон. allowedContextTags — теги, которые в этом месте конфига
    /// действительно кто-то подставит (см. ContextualTags).
    public static IReadOnlyList<TemplateIssue> Analyze(
        string? template,
        Config config,
        string where,
        IReadOnlyCollection<string>? allowedContextTags = null)
    {
        var issues = new List<TemplateIssue>();
        if (string.IsNullOrEmpty(template)) return issues;

        var allowed = allowedContextTags == null
            ? new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            : new HashSet<string>(allowedContextTags, StringComparer.OrdinalIgnoreCase);

        Walk(template, config, where, allowed, issues,
            new HashSet<string>(StringComparer.OrdinalIgnoreCase), 0);

        return issues;
    }

    /// Полнота переводов: ключ есть, но не на всех языках, которые встречаются в Messages.json.
    /// Отдельным проходом, а не внутри Walk: иначе одна и та же претензия повторялась бы
    /// столько раз, сколько шаблонов ссылается на ключ.
    public static IReadOnlyList<TemplateIssue> AnalyzeLanguageCoverage(Config config)
    {
        var issues = new List<TemplateIssue>();
        var languages = CollectLanguages(config);
        if (config.LanguageMessages == null || languages.Count == 0) return issues;

        foreach (var (key, translations) in config.LanguageMessages)
        {
            if (translations == null) continue;

            foreach (var lang in languages)
            {
                if (translations.ContainsKey(lang)) continue;

                issues.Add(new TemplateIssue(
                    TemplateSeverity.Warning,
                    $"Messages.json → {key}",
                    "{" + key + "}",
                    $"нет перевода на «{lang}» — игрок с этим языком увидит текст на другом языке"));
            }
        }

        return issues;
    }

    /// Все языки, встречающиеся в конфиге: объединение ключей переводов плюс DefaultLang.
    internal static IReadOnlyList<string> CollectLanguages(Config config)
    {
        var languages = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);

        if (!string.IsNullOrEmpty(config.DefaultLang)) languages.Add(config.DefaultLang);

        if (config.LanguageMessages != null)
        {
            foreach (var translations in config.LanguageMessages.Values)
            {
                if (translations == null) continue;
                foreach (var lang in translations.Keys) languages.Add(lang);
            }
        }

        var result = new List<string>(languages.Count);
        result.AddRange(languages);
        return result;
    }

    private static void Walk(
        string text,
        Config config,
        string where,
        HashSet<string> allowedContextTags,
        List<TemplateIssue> issues,
        HashSet<string> expandedKeys,
        int depth)
    {
        if (depth > MaxDepth) return;

        foreach (Match match in MessageProcessor.TagPattern.Matches(text))
        {
            var tag = match.Groups[0].Value;
            var name = match.Groups[1].Value;

            if (TextFormatter.KnownColorTags.Contains(tag)) continue;
            if (MessageProcessor.IsSystemTag(tag)) continue;
            if (allowedContextTags.Contains(tag)) continue;

            if (ContextualTags.TryGetValue(tag, out var worksIn))
            {
                issues.Add(new TemplateIssue(TemplateSeverity.Warning, where, tag,
                    $"тег {tag} здесь никто не подставит — он работает только в: {worksIn}"));
                continue;
            }

            if (config.LanguageMessages != null &&
                config.LanguageMessages.TryGetValue(name, out var translations))
            {
                // Ключ известен. Его тексты сами содержат теги — проверяем и их.
                if (translations == null || !expandedKeys.Add(name)) continue;

                foreach (var (lang, value) in translations)
                {
                    if (string.IsNullOrEmpty(value)) continue;
                    Walk(value, config, $"{where} → {tag}[{lang}]", allowedContextTags, issues,
                        expandedKeys, depth + 1);
                }

                continue;
            }

            issues.Add(new TemplateIssue(TemplateSeverity.Error, where, tag,
                $"неизвестный тег {tag} — игрок увидит его как текст. " +
                "Добавьте ключ в Messages.json или исправьте опечатку"));
        }
    }
}

File: metamod/src/core/template_diagnostics.cpp
#include "core/template_diagnostics.h"

#include "core/message_processor.h"
#include "core/text_formatter.h"

#include <set>
#include <utility>

namespace nm {

namespace context {
const std::vector<std::string> kPlayerNameOnly = {"{PLAYERNAME}"};
const std::vector<std::string> kRestart = {"{TIME_RESTART}", "{SECONDS}"};
const std::vector<std::string> kChangeTeam = {"{PLAYERNAME}", "{TEAM}", "{OLD_TEAM}"};
const std::vector<std::string> kJoinTeam = {"{PLAYERNAME}", "{TEAM}"};
const std::vector<std::string> kJoinLeave = {"{PLAYERNAME}", "{COUNTRY}", "{CITY}"};
const std::vector<std::string> kServer = {"{SERVER_IP}", "{SERVER_PORT}", "{SERVER_MAP}",
                                          "{SERVER_PLAYERS}", "{SERVER_MAXPLAYERS}"};
}  // namespace context

namespace {

// Больше трёх уровней вложенности ({prefix} -> текст -> ещё ключ) не бывает,
// а предел заодно страхует от циклов {a} -> {b} -> {a}
constexpr int kMaxDepth = 3;

using CiSet = std::set<std::string, CaseInsensitiveLess>;

// Теги, которые подставляет вызывающий код, а не MessageProcessor.
// Значение — человекочитаемое «где именно».
const CiMap<std::string>& ContextualTags() {
    static const CiMap<std::string> tags = {
        {"{PLAYERNAME}", "WelcomeMessage, ChangeTeamMessage, JoinTeamMessage, Join/LeaveMessages"},
        {"{TEAM}", "ChangeTeamMessage, JoinTeamMessage"},
        {"{OLD_TEAM}", "ChangeTeamMessage"},
        {"{TIME_RESTART}", "RestartNotify"},
        {"{SECONDS}", "RestartNotify"},
        {"{COUNTRY}", "JoinMessages, LeaveMessages"},
        {"{CITY}", "JoinMessages, LeaveMessages"},
        {"{SERVER_IP}", "Servers.json"},
        {"{SERVER_PORT}", "Servers.json"},
        {"{SERVER_MAP}", "Servers.json"},
        {"{SERVER_PLAYERS}", "Servers.json"},
        {"{SERVER_MAXPLAYERS}", "Servers.json"},
    };
    return tags;
}

void Walk(const std::string& text, const Config& config, const std::string& where,
          const CiSet& allowed, std::vector<TemplateIssue>* issues, CiSet* expanded, int depth) {
    if (depth > kMaxDepth) return;

    for (const TagMatch& match : MessageProcessor::FindTags(text)) {
        if (text::IsKnownColorTag(match.tag)) continue;
        if (MessageProcessor::IsSystemTag(match.tag)) continue;
        if (allowed.count(match.tag) != 0) continue;

        const auto contextual = ContextualTags().find(match.tag);
        if (contextual != ContextualTags().end()) {
            issues->push_back({TemplateSeverity::Warning, where, match.tag,
                               "тег " + match.tag + " здесь никто не подставит — он работает только в: " +
                                   contextual->second});
            continue;
        }

        const auto translations = config.languageMessages.find(match.name);
        if (translations != config.languageMessages.end()) {
            // Ключ известен. Его тексты сами содержат теги — проверяем и их
            if (!expanded->insert(match.name).second) continue;
            for (const auto& lang : translations->second) {
                if (lang.second.empty()) continue;
                Walk(lang.second, config, where + " → " + match.tag + "[" + lang.first + "]", allowed,
                     issues, expanded, depth + 1);
            }
            continue;
        }

        issues->push_back({TemplateSeverity::Error, where, match.tag,
                           "неизвестный тег " + match.tag +
                               " — игрок увидит его как текст. Добавьте ключ в Messages.json "
                               "или исправьте опечатку"});
    }
}

void Check(const std::string& templateText, const Config& config, const std::string& where,
           const std::vector<std::string>& context, std::vector<TemplateIssue>* issues) {
    for (TemplateIssue& issue : AnalyzeTemplate(templateText, config, where, context)) {
        issues->push_back(std::move(issue));
    }
}

void CheckMessageList(const CiMap<std::vector<std::string>>& messages, const std::string& where,
                      const Config& config, std::vector<TemplateIssue>* issues) {
    for (const auto& lang : messages) {
        for (size_t i = 0; i < lang.second.size(); ++i) {
            Check(lang.second[i], config, where + "[" + lang.first + "] #" + std::to_string(i + 1),
                  context::kJoinLeave, issues);
        }
    }
}

}  // namespace

std::string TemplateIssue::ToString() const {
    return std::string(severity == TemplateSeverity::Error ? "[ОШИБКА] " : "[внимание] ") + where +
           ": " + text;
}

std::vector<TemplateIssue> AnalyzeTemplate(const std::string& templateText, const Config& config,
                                           const std::string& where,
                                           const std::vector<std::string>& allowedContextTags) {
    std::vector<TemplateIssue> issues;
    if (templateText.empty()) return issues;

    const CiSet allowed(allowedContextTags.begin(), allowedContextTags.end());
    CiSet expanded;
    Walk(templateText, config, where, allowed, &issues, &expanded, 0);
    return issues;
}

std::vector<TemplateIssue> AnalyzeLanguageCoverage(const Config& config) {
    std::vector<TemplateIssue> issues;

    // Все языки конфига: ключи переводов плюс DefaultLang
    CiSet languages;
    if (!config.defaultLang.empty()) languages.insert(config.defaultLang);
    for (const auto& key : config.languageMessages) {
        for (const auto& lang : key.second) languages.insert(lang.first);
    }
    if (languages.empty()) return issues;

    for (const auto& key : config.languageMessages) {
        for (const std::string& lang : languages) {
            if (key.second.count(lang) != 0) continue;
            issues.push_back({TemplateSeverity::Warning, "Messages.json → " + key.first, "{" + key.first + "}",
                              "нет перевода на «" + lang +
                                  "» — игрок с этим языком увидит текст на другом языке"});
        }
    }
    return issues;
}

std::vector<TemplateIssue> CollectIssues(const Config& config) {
    std::vector<TemplateIssue> issues;
    const std::vector<std::string> none;

    // --- Settings.json ---
    if (config.welcomeMessage) {
        Check(config.welcomeMessage->message, config, "Settings.json → WelcomeMessage",
              context::kPlayerNameOnly, &issues);
    }
    Check(config.changeTeamMessage, config, "Settings.json → ChangeTeamMessage", context::kChangeTeam, &issues);
    Check(config.joinTeamMessage, config, "Settings.json → JoinTeamMessage", context::kJoinTeam, &issues);
    Check(config.titleAnnounceServers, config, "Settings.json → TitleAnnounceServers", none, &issues);

    if (config.restartNotify) {
        Check(config.restartNotify->defaultMessage, config, "Settings.json → RestartNotify.DefaultMessage",
              context::kRestart, &issues);
        for (const auto& threshold : config.restartNotify->thresholds) {
            Check(threshold.second, config,
                  "Settings.json → RestartNotify.Thresholds[" + threshold.first + "]", context::kRestart,
                  &issues);
        }
    }

    // --- Messages.json ---
    CheckMessageList(config.joinMessages, "Messages.json → JoinMessages", config, &issues);
    CheckMessageList(config.leaveMessages, "Messages.json → LeaveMessages", config, &issues);
    for (TemplateIssue& issue : AnalyzeLanguageCoverage(config)) issues.push_back(std::move(issue));

    // --- Ads.json ---
    for (size_t i = 0; i < config.ads.size(); ++i) {
        for (const OrderedPairs& block : config.ads[i].messages) {
            for (const auto& channel : block) {
                Check(channel.second, config,
                      "Ads.json → блок #" + std::to_string(i + 1) + ", " + channel.first, none, &issues);
            }
        }
    }

    // --- Servers.json ---
    if (config.servers) {
        for (size_t i = 0; i < config.servers->list.size(); ++i) {
            const ServerData& server = config.servers->list[i];
            const std::string index = std::to_string(i + 1);
            Check(server.messageTemplate, config, "Servers.json → сервер #" + index + ", MessageTemplate",
                  context::kServer, &issues);
            Check(server.messageTemplateConsole, config,
                  "Servers.json → сервер #" + index + ", MessageTemplateConsole", context::kServer, &issues);
        }
    }

    return issues;
}

}  // namespace nm
